### ISSUE TITLE
[no-var][semi] behaviour for interfaces and declare var

### ISSUE THREAD

**Issue by ficristo** [bradzacher/eslint-plugin-typescript#266]
**Repro**
See below.

```JSON
{
  "rules": {
    "no-var": "error",
    "semi": "error"
  }
}
```

```TS
interface Object {
    foo(): string,
}
declare var Object: {
    new (value?: any): Object,
    foo(): string
}
```
```TS
declare namespace CodeMirror {
    export var Doc : CodeMirror.DocConstructor;
    var version: string;
}
```

**Expected Result**
No `no-var` errors.
`semi` errors for the three lines inside the curly braces. (I'm not sure if commas are valid, it seems typescript is quite lenient about them). No errors after the curly braces.

**Actual Result**
`no-var` errors.
no `semi` errors but after the last curly brace.

**Additional Info**

**Versions**

| package                    | version |
| -------------------------- | ------- |
| `eslint-plugin-typescript` | `1.0.0-rc.0` |
| `typescript-eslint-parser` | `eslint-plugin-typescript/parser` |
| `typescript`               | `3.1.6` |

**Comment by bradzacher** [comment]
```TS
declare namespace CodeMirror {
    export var Doc : CodeMirror.DocConstructor;
    var version: string; // error: 'version' is defined but never used.
}
```

`var`s are automatically exported in typescript namespaces ([playground](https://www.typescriptlang.org/play/index.html#src=declare%20namespace%20CodeMirror%20%7B%0A%20%20%20%20export%20var%20Doc%20%3A%20string%3B%0A%20%20%20%20var%20version%3A%20string%3B%0A%7D%0A%0ACodeMirror.version.big()%3B%0A)), so version should be marked as defined. This is an easy fix as we can just check the type of the parent.

```TS
interface Object {
    foo(): string,
}
declare var Object: { // error: 'Object' is defined but never used.
    new (value?: any): Object,
    foo(): string
}
```

ambient variable declarations should be marked as used automatically.
Unfortunately the parser doesn't give us a flag for if a variable is marked as `declare` or not ([ast explorer](https://astexplorer.net/#/gist/48223ee9341fd8015f8200e29de2ad1e/cabea38a1d3ed4727dbc44b3cfb01bb8c771c72c)).

Will need to get this added.
This is a low-priority use case though, as not many people use global ambient definitions.
So we won't block 1.0.0 on this.

**Comment by bradzacher** [comment]
As for your `semi` desired errors;

```TS
interface Object {
    foo(): string,
}
declare var Object: {
    new (value?: any): Object,
    foo(): string
} // error: Missing semicolon.
```
use [`typescript/member-delimiter-style`](https://github.com/bradzacher/eslint-plugin-typescript/blob/master/docs/rules/member-delimiter-style.md). This will enforce styling for your type/interface definitions.

`semi` should (and does) only error on the missing semicolon for the variable declaration.

---

```TS
declare namespace CodeMirror {
    export var Doc : CodeMirror.DocConstructor;
    var version: string;
}
declare namespace CodeMirror {
    export var Doc : CodeMirror.DocConstructor // error: Missing semicolon.
    var version: string // error: Missing semicolon.
}
```

`semi` correctly handles variable definitions within a `namespace`.

**Comment by armano2** [comment]
some time ago i added support for declare: https://github.com/JamesHenry/typescript-estree/pull/42
but we are still waiting for https://github.com/eslint/typescript-eslint-parser/pull/584 to be merged before we can use it here

**Comment by bradzacher** [comment]
My earlier comment was about `no-unused-vars`, when this issue was about `no-var`

I would say that the errors are correct.

```TS
declare namespace CodeMirror {
    export var Doc : CodeMirror.DocConstructor; // error: Unexpected var, use let or const instead.
    var version: string; // error: Unexpected var, use let or const instead.
}
```

You should instead use `let`/`const` in your namespace declarations ([playground](https://www.typescriptlang.org/play/index.html#src=declare%20namespace%20CodeMirror%20%7B%0A%20%20%20%20export%20let%20Doc%20%3A%20string%3B%0A%20%20%20%20const%20version%3A%20string%3B%0A%7D%0A%0ACodeMirror.Doc%20%3D%20''%3B%0ACodeMirror.version%20%3D%20''%3B%0A)).

---

```TS
interface Object {
    foo(): string,
}
declare var Object: { // error: Unexpected var, use let or const instead.
    new (value?: any): Object,
    foo(): string
}
```

It's hard to say without knowing your use case; but I think that this is probably incorrect use of `declare var`.
Extending the base `Object` definition is a weird thing to do, and would have unintended side-effects.
A better solution would be to use a `.d.ts` file and [extend the definition](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c47ee9bf275423ee40b3f4101888f6e461f2c18a/types/es6-shim/ts3.1/index.d.ts#L300-L323).

Though we could add support for skipping `declare`d `var`s.
It's a bit hard for us though, as we'd have to redefine the entire rule to make it work.

**Comment by ficristo** [comment]
My mistake: the `semi` errors were about only the first snippet, the second one, as you noted, is correct.
I'll take a look to `typescript/member-delimiter-style`.
For now I'm fine to know I should use `let` \ `const` and avoid a `declare var`.
Thank you.
